src/NuGet.Services.AzureSearch/SearchService/SecretRefresher.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using NuGet.Services.AzureSearch.Wrappers;
+using NuGet.Services.KeyVault;
+
+namespace NuGet.Services.AzureSearch.SearchService
+{
+    public class SecretRefresher : ISecretRefresher
+    {
+        private readonly IRefreshableSecretReaderFactory _factory;
+        private readonly ISystemTime _systemTime;
+        private readonly IOptionsSnapshot<SearchServiceConfiguration> _options;
+        private readonly ILogger<SecretRefresher> _logger;
+
+        public SecretRefresher(
+            IRefreshableSecretReaderFactory factory,
+            ISystemTime systemTime,
+            IOptionsSnapshot<SearchServiceConfiguration> options,
+            ILogger<SecretRefresher> logger)
+        {
+            _factory = factory ?? throw new ArgumentNullException(nameof(factory));
+            _systemTime = systemTime ?? throw new ArgumentNullException(nameof(systemTime));
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public async Task RefreshContinuouslyAsync(CancellationToken token)
+        {
+            while (!token.IsCancellationRequested)
+            {
+                _logger.LogInformation("Trying to refresh the secrets.");
+
+                TimeSpan delay;
+                try
+                {
+                    await _factory.RefreshAsync(token);
+                    delay = _options.Value.SecretRefreshFrequency;
+                }
+                catch (Exception ex)
+                {
+                    _logger.LogError(0, ex, "An exception was thrown while refreshing the secrets.");
+
+                    delay = _options.Value.SecretRefreshFailureRetryFrequency;
+                }
+
+                if (token.IsCancellationRequested)
+                {
+                    return;
```

Comment:
This state seems to be an error state because the refresh action should not be cancelled. Should it be monitored?

Reply:
This token is returned by the background worker system. It will be cancelled when the app is shutting down.